cip/CIP2015-05-13-EXISTS.adoc

```diff
@@ -0,0 +1,174 @@
+= CIP2015-05-13 - EXISTENTIAL SUBQUERIES
+:numbered:
+:toc:
+:toc-placement: macro
+:source-highlighter: codemirror
+
+*Authors:* Andrés Taylor <andres@neotechnology.com>
+
+
+[abstract]
+.Abstract
+--
+This CIP introduces EXISTS, a general function that can test for the existence of properties, simple patterns and full subqueries.
+--
+
+toc::[]
+
+== Background
+
+In a way, Cypher already has existential subqueries - pattern predicates allow checking for the existence of subgraphs.
+This form is helpful, but not exhaustive in the types of subqueries that users want to be able to express.
+
+== Proposal
+
+To make this feature more powerful, add a new function named +EXISTS+, that would exist in two forms:
+
+ * Property existence checking
+ * Subquery existence checking
+
+=== Syntax
+
+----
+expression = …
+           | property exists
+           | subquery exists
+           | simple subquery exists
+            ;
+
+property exists = "EXISTS", "(", expression ")"
+                ;
+
+subquery exists = "EXISTS", "{", read only clause, { read only clause }, "}" ;
+
+simple subquery exists = "EXISTS", "{", simple match, "}" ;
+
+simple match = pattern, { ",", pattern }, [ "WHERE", predicate ] ;
+
+----
+Note: the read-only clause above includes WITH but not RETURN.
+
+=== Semantics
+
+All forms of +EXIST(...)+ accomplish the same task - check if a particular pattern exists in the database.
+They are expressions, and as such must be side-effect free; that is, the subqueries in +EXIST+ must not be updating queries.
+Regarding nomenclature, all forms of +EXISTS+ are scalar expressions - they return a single boolean value, and this value is never +null+.
+
+==== Property exists
+
+In the property checking form, any expression is accepted as the input.
+If the input expression returns a +null+ value, +EXISTS(...)+ returns false, and for all other values, it will return +true+.
+This form of +EXISTS(...)+ is equivalent to checking for +null+ using +IS NOT NULL+; i.e. +EXISTS(expr)+ is equivalent to +expr IS NOT NULL+.
```

Comment:
Nit picking: I would phrase this slightly differently.

    If the input expression would have produced a +null+ value, +EXISTS(...)+ will produce +false+, for all other values it will produce +true+.

1. it doesn't actually return (unless the expression occurs in the `RETURN` clause)
2. I prefer to say _would produce_ rather than _produces_, since the actual implementation might execute the `EXISTS` in a different way than actually evaluate the input expression and checking it for `null`.